File: bench/packet_classifier.c
// benchmark-contract: branch-lcg32;seed=123456789;iterations=25000000;threshold=2147483648
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

static inline void emit_checksum(uint64_t value) {
  // Every peer prints this one line and nothing else; the 63-bit mask
  // keeps the value printable by the languages without unsigned 64-bit
  // integers (NURL's `i`, Python's signed int, JS BigInt) so the five
  // outputs can be compared byte for byte.
  printf("%llu\n", (unsigned long long)(value & 0x7fffffffffffffffULL));
}

int main(void) {
  const uint64_t iterations = 25000000ULL;
  const uint64_t threshold = 1ULL << 31;
  uint64_t state = 123456789ULL;
  uint64_t i = 0;

  while (i < iterations) {
    if (state < threshold) {
      state = (state * 1664525ULL + 1013904223ULL) & 0xffffffffULL;
    } else {
      state = (state * 22695477ULL + 1ULL) & 0xffffffffULL;
    }
    ++i;
  }

  emit_checksum(state);
  return 0;
}

File: bench/packet_classifier.rs
// benchmark-contract: branch-lcg32;seed=123456789;iterations=25000000;threshold=2147483648
const ITERATIONS: u64 = 25_000_000;
const THRESHOLD: u64 = 1u64 << 31;
const SEED: u64 = 123_456_789;
const MASK: u64 = 0xffff_ffff;

fn finish(value: u64) -> ! {
    // See the C peer: one decimal line, masked to 63 bits.
    println!("{}", value & 0x7fff_ffff_ffff_ffff);
    std::process::exit(0)
}

fn main() {
    let mut state = SEED;
    let mut i = 0u64;
    while i < ITERATIONS {
        if state < THRESHOLD {
            state = state
                .wrapping_mul(1_664_525)
                .wrapping_add(1_013_904_223)
                & MASK;
        } else {
            state = state.wrapping_mul(22_695_477).wrapping_add(1) & MASK;
        }
        i += 1;
    }
    finish(state);
}
